File: apps/worker/src/routes/setting.ts
import type { DBDump, ImportResult } from "@wheel/core"
import type { AppEnv } from "../runtime/types"
import { Hono } from "hono"
import { getAllSettings, updateSettings } from "../db/dal/settings"
import * as schema from "../db/schema"

const DEFAULT_SETTINGS: Record<string, string> = {
  log_retention_days: "30",
  circuit_breaker_threshold: "5",
  circuit_breaker_cooldown: "60",
  circuit_breaker_max_cooldown: "600",
}

const settingRoutes = new Hono<AppEnv>()

settingRoutes.get("/", async (c) => {
  const db = c.env.DB
  const settings = await getAllSettings(db)
  // Merge defaults for keys that don't exist yet, only return configurable keys
  const merged = { ...DEFAULT_SETTINGS }
  for (const key of Object.keys(DEFAULT_SETTINGS)) {
    if (key in settings) {
      merged[key] = settings[key]
    }
  }
  return c.json({ success: true, data: { settings: merged } })
})

settingRoutes.post("/update", async (c) => {
  const body = await c.req.json()
  const db = c.env.DB
  await updateSettings(db, body.settings)
  return c.json({ success: true })
})

// Export all data as JSON file
settingRoutes.get("/export", async (c) => {
  try {
    const includeLogs = c.req.query("include_logs") === "true"
    const db = c.env.DB

    const [allChannels, allChannelKeys, allGroups, allGroupItems, allApiKeys, allSettings] =
      await Promise.all([
        db.select().from(schema.channels),
        db.select().from(schema.channelKeys),
        db.select().from(schema.groups),
        db.select().from(schema.groupItems),
        db.select().from(schema.apiKeys),
        db.select().from(schema.settings),
      ])

    // Assemble channels with their keys (matching the Channel model shape)
    const channelsWithKeys = allChannels.map((ch) => ({
      ...ch,
      keys: allChannelKeys.filter((k) => k.channelId === ch.id),
    }))

    // Assemble groups with their items (matching the Group model shape)
    const groupsWithItems = allGroups.map((g) => ({
      ...g,
      items: allGroupItems.filter((i) => i.groupId === g.id),
    }))

    const dump: DBDump = {
      version: 1,
      exportedAt: new Date().toISOString(),
      channels: channelsWithKeys,
      groups: groupsWithItems,
      apiKeys: allApiKeys,
      settings: allSettings,
    }

    if (includeLogs) {
      dump.relayLogs = await db.select().from(schema.relayLogs)
    }

    return new Response(JSON.stringify(dump, null, 2), {
      headers: {
        "Content-Type": "application/json",
        "Content-Disposition": `attachment; filename="wheel-export-${Date.now()}.json"`,
      },
    })
  } catch (e) {
    const message = e instanceof Error ? e.message : "Export failed"
    return c.json({ success: false, error: message }, 500)
  }
})

// Import data from JSON
settingRoutes.post("/import", async (c) => {
  try {
    let dump: DBDump

    const contentType = c.req.header("content-type") || ""
    if (contentType.includes("multipart/form-data")) {
      const formData = await c.req.formData()
      const file = formData.get("file") as File | null
      if (!file) {
        return c.json({ success: false, error: "No file provided" }, 400)
      }
      const text = await file.text()
      dump = JSON.parse(text)
    } else {
      dump = await c.req.json()
    }

    if (!dump.version || !dump.exportedAt) {
      return c.json({ success: false, error: "Invalid dump format" }, 400)
    }

    const db = c.env.DB

    const result: ImportResult = {
      channels: { added: 0, skipped: 0 },
      groups: { added: 0, skipped: 0 },
      apiKeys: { added: 0, skipped: 0 },
      settings: { added: 0, skipped: 0 },
    }

    // Import channels (match by name)
    if (dump.channels?.length) {
      const existingChannels = await db.select().from(schema.channels)
      const existingNames = new Set(existingChannels.map((ch) => ch.name))

      for (const ch of dump.channels) {
        if (existingNames.has(ch.name)) {
          result.channels.skipped++
          continue
        }

        const { id, keys, ...channelData } = ch as typeof ch & { keys?: unknown[] }
        const [inserted] = await db.insert(schema.channels).values(channelData).returning()

        // Import channel keys
        if (keys && Array.isArray(keys) && keys.length > 0) {
          await db.insert(schema.channelKeys).values(
            keys.map((k: any) => ({
              channelId: inserted.id,
              enabled: k.enabled ?? true,
              channelKey: k.channelKey,
              statusCode: k.statusCode ?? 0,
              lastUseTimestamp: k.lastUseTimestamp ?? 0,
              totalCost: k.totalCost ?? 0,
              remark: k.remark ?? "",
            })),
          )
        }

        result.channels.added++
      }
    }

    // Import groups (match by name)
    if (dump.groups?.length) {
      const existingGroups = await db.select().from(schema.groups)
      const existingNames = new Set(existingGroups.map((g) => g.name))

      // Build a name-to-id mapping for channels (needed for groupItems)
      const allChannels = await db.select().from(schema.channels)
      const channelNameToId = new Map(allChannels.map((ch) => [ch.name, ch.id]))

      // Also build old-id to name mapping from the dump
      const dumpChannelIdToName = new Map((dump.channels || []).map((ch) => [ch.id, ch.name]))

      for (const g of dump.groups) {
        if (existingNames.has(g.name)) {
          result.groups.skipped++
          continue
        }

        const { id, items, ...groupData } = g as typeof g & { items?: unknown[] }
        const [inserted] = await db.insert(schema.groups).values(groupData).returning()

        // Import group items, remapping channelId from dump to current DB
        if (items && Array.isArray(items) && items.length > 0) {
          const validItems: (typeof schema.groupItems.$inferInsert)[] = []
          for (const item of items as any[]) {
            // Resolve channel ID: find the channel name from the dump, then find its current ID
            const channelName = dumpChannelIdToName.get(item.channelId)
            const currentChannelId = channelName ? channelNameToId.get(channelName) : undefined

            if (currentChannelId) {
              validItems.push({
                groupId: inserted.id,
                channelId: currentChannelId,
                modelName: item.modelName ?? "",
                priority: item.priority ?? 0,
                weight: item.weight ?? 1,
              })
            }
          }

          if (validItems.length > 0) {
            await db.insert(schema.groupItems).values(validItems)
          }
        }

        result.groups.added++
      }
    }

    // Import API keys (match by apiKey value)
    if (dump.apiKeys?.length) {
      const existingKeys = await db.select().from(schema.apiKeys)
      const existingKeyValues = new Set(existingKeys.map((k) => k.apiKey))

      for (const ak of dump.apiKeys) {
        if (existingKeyValues.has(ak.apiKey)) {
          result.apiKeys.skipped++
          continue
        }

        const { id, ...keyData } = ak
        await db.insert(schema.apiKeys).values(keyData)
        result.apiKeys.added++
      }
    }

    // Import settings (match by key, skip existing)
    if (dump.settings?.length) {
      const existingSettings = await db.select().from(schema.settings)
      const existingKeySet = new Set(existingSettings.map((s) => s.key))

      for (const s of dump.settings) {
        if (existingKeySet.has(s.key)) {
          result.settings.skipped++
          continue
        }

        await db.insert(schema.settings).values({ key: s.key, value: s.value })
        result.settings.added++
      }
    }

    // Invalidate caches
    await Promise.all([
      c.env.CACHE.delete("channels"),
      c.env.CACHE.delete("groups"),
      c.env.CACHE.delete("apikeys"),
      c.env.CACHE.delete("settings"),
    ])

    return c.json({ success: true, data: result })
  } catch (e) {
    const message = e instanceof Error ? e.message : "Import failed"
    return c.json({ success: false, error: message }, 500)
  }
})

export { settingRoutes }

File: apps/worker/internal/handler/setting.go
package handler

import (
	"bufio"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/kunish/wheel/apps/worker/internal/config"
	"github.com/kunish/wheel/apps/worker/internal/db/dal"
	"github.com/kunish/wheel/apps/worker/internal/service"
	"github.com/kunish/wheel/apps/worker/internal/types"
)

// Default settings values matching the TS implementation.
var defaultSettings = map[string]string{
	"log_retention_days":           "365",
	"circuit_breaker_threshold":    "5",
	"circuit_breaker_cooldown":     "60",
	"circuit_breaker_max_cooldown": "600",
	"active_profile_id":            "0",
}

// ──── Setting Routes ────

// GetSettings godoc
// @Summary Get all settings
// @Tags Settings
// @Produce json
// @Success 200 {object} object "{success: true, data: {settings: map[string]string}}"
// @Security BearerAuth
// @Router /api/v1/setting/ [get]
func (h *Handler) GetSettings(c *gin.Context) {
	settings, err := dal.GetAllSettings(c.Request.Context(), h.DB)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}

	merged := make(map[string]string)
	for k, v := range defaultSettings {
		merged[k] = v
	}
	for k := range defaultSettings {
		if v, ok := settings[k]; ok {
			merged[k] = v
		}
	}

	successJSON(c, gin.H{"settings": merged})
}

// UpdateSettings godoc
// @Summary Update settings
// @Tags Settings
// @Accept json
// @Produce json
// @Param body body types.SettingsUpdateRequest true "Settings key-value pairs"
// @Success 200 {object} object "{success: true}"
// @Failure 400 {object} object "{success: false, error: string}"
// @Security BearerAuth
// @Router /api/v1/setting/update [post]
func (h *Handler) UpdateSettings(c *gin.Context) {
	var req types.SettingsUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		errorJSON(c, http.StatusBadRequest, "Invalid request body")
		return
	}

	if err := dal.UpdateSettings(c.Request.Context(), h.DB, req.Settings); err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}

	successNoData(c)
}

// ExportData godoc
// @Summary Export all data as JSON file
// @Tags Settings
// @Produce json
// @Param include_logs query string false "Include relay logs (true|false)"
// @Success 200 {file} file "JSON export file"
// @Security BearerAuth
// @Router /api/v1/setting/export [get]
func (h *Handler) ExportData(c *gin.Context) {
	includeLogs := c.Query("include_logs") == "true"
	ctx := c.Request.Context()

	channels, err := dal.ListChannels(ctx, h.DB)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}
	groups, err := dal.ListGroups(ctx, h.DB, 0)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}
	apiKeys, err := dal.ListApiKeys(ctx, h.DB)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}
	settings, err := dal.GetAllSettings(ctx, h.DB)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, err.Error())
		return
	}

	settingsList := make([]types.Setting, 0, len(settings))
	for k, v := range settings {
		settingsList = append(settingsList, types.Setting{Key: k, Value: v})
	}

	dump := types.DBDump{
		Version:    1,
		ExportedAt: time.Now().UTC().Format(time.RFC3339),
		Channels:   channels,
		Groups:     groups,
		APIKeys:    apiKeys,
		Settings:   settingsList,
	}

	if includeLogs {
		logs, _, err := dal.ListLogs(ctx, h.DB, dal.ListLogsOpts{Page: 1, PageSize: 999999})
		if err != nil {
			errorJSON(c, http.StatusInternalServerError, err.Error())
			return
		}
		dump.RelayLogs = logs
	}

	data, err := json.MarshalIndent(dump, "", "  ")
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, "Failed to marshal export data")
		return
	}

	c.Header("Content-Type", "application/json")
	c.Header("Content-Disposition", fmt.Sprintf(`attachment; filename="wheel-export-%d.json"`, time.Now().UnixMilli()))
	c.Data(http.StatusOK, "application/json", data)
}

// ImportData godoc
// @Summary Import data from JSON file or body
// @Tags Settings
// @Accept json,mpfd
// @Produce json
// @Param file formData file false "JSON export file"
// @Success 200 {object} object "{success: true, data: ImportResult}"
// @Failure 400 {object} object "{success: false, error: string}"
// @Security BearerAuth
// @Router /api/v1/setting/import [post]
func (h *Handler) ImportData(c *gin.Context) {
	var dump types.DBDump

	contentType := c.GetHeader("Content-Type")
	if strings.Contains(contentType, "multipart/form-data") {
		file, _, err := c.Request.FormFile("file")
		if err != nil {
			errorJSON(c, http.StatusBadRequest, "No file provided")
			return
		}
		defer file.Close()
		if err := json.NewDecoder(file).Decode(&dump); err != nil {
			errorJSON(c, http.StatusBadRequest, "Invalid JSON in file")
			return
		}
	} else {
		if err := c.ShouldBindJSON(&dump); err != nil {
			errorJSON(c, http.StatusBadRequest, "Invalid request body")
			return
		}
	}

	if dump.Version == 0 || dump.ExportedAt == "" {
		errorJSON(c, http.StatusBadRequest, "Invalid dump format")
		return
	}

	result := service.ImportData(c.Request.Context(), h.DB, &dump)

	// Invalidate caches
	h.Cache.Delete("channels")
	h.Cache.Delete("groups")
	h.Cache.Delete("apikeys")
	h.Cache.Delete("settings")

	successJSON(c, result)
}

// ResetCircuitBreakers godoc
// @Summary Reset all circuit breakers
// @Tags Settings
// @Produce json
// @Success 200 {object} object "{success: true, data: {reset: int}}"
// @Security BearerAuth
// @Router /api/v1/setting/reset-circuit-breakers [post]
func (h *Handler) ResetCircuitBreakers(c *gin.Context) {
	if h.CircuitBreakers == nil {
		errorJSON(c, http.StatusInternalServerError, "Circuit breaker manager not available")
		return
	}
	count := h.CircuitBreakers.ResetAll()
	successJSON(c, gin.H{"reset": count})
}

// GetVersion godoc
// @Summary Get current server version
// @Tags Settings
// @Produce json
// @Success 200 {object} object "{success: true, data: {version: string}}"
// @Security BearerAuth
// @Router /api/v1/setting/version [get]
func (h *Handler) GetVersion(c *gin.Context) {
	successJSON(c, gin.H{"version": strings.TrimPrefix(config.Version, "v")})
}

// CheckUpdate godoc
// @Summary Check for new releases on GitHub
// @Tags Settings
// @Produce json
// @Success 200 {object} object "{success: true, data: {current, latest, hasUpdate, releaseUrl, releaseNotes}}"
// @Security BearerAuth
// @Router /api/v1/setting/check-update [get]
func (h *Handler) CheckUpdate(c *gin.Context) {
	current := strings.TrimPrefix(config.Version, "v")

	resp, err := http.Get("https://api.github.com/repos/kunish/wheel/releases/latest")
	if err != nil {
		errorJSON(c, http.StatusBadGateway, "Failed to reach GitHub API")
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
		errorJSON(c, http.StatusBadGateway, fmt.Sprintf("GitHub API returned %d: %s", resp.StatusCode, string(body)))
		return
	}

	var release struct {
		TagName string `json:"tag_name"`
		HTMLURL string `json:"html_url"`
		Body    string `json:"body"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		errorJSON(c, http.StatusBadGateway, "Failed to parse GitHub response")
		return
	}

	latest := strings.TrimPrefix(release.TagName, "v")

	successJSON(c, gin.H{
		"current":      current,
		"latest":       latest,
		"hasUpdate":    latest != current,
		"releaseUrl":   release.HTMLURL,
		"releaseNotes": release.Body,
	})
}

// ApplyUpdate godoc
// @Summary Download latest release binary and restart
// @Tags Settings
// @Produce json
// @Success 200 {object} object "{success: true}"
// @Failure 502 {object} object "{success: false, error: string}"
// @Security BearerAuth
// @Router /api/v1/setting/apply-update [post]
func (h *Handler) ApplyUpdate(c *gin.Context) {
	// 1. Fetch latest release info (with assets)
	resp, err := http.Get("https://api.github.com/repos/kunish/wheel/releases/latest")
	if err != nil {
		errorJSON(c, http.StatusBadGateway, "Failed to reach GitHub API")
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		errorJSON(c, http.StatusBadGateway, fmt.Sprintf("GitHub API returned %d", resp.StatusCode))
		return
	}

	var release struct {
		TagName string `json:"tag_name"`
		Assets  []struct {
			Name               string `json:"name"`
			BrowserDownloadURL string `json:"browser_download_url"`
		} `json:"assets"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		errorJSON(c, http.StatusBadGateway, "Failed to parse GitHub response")
		return
	}

	// 2. Find matching binary asset and checksums asset
	wantName := fmt.Sprintf("wheel-%s-%s", runtime.GOOS, runtime.GOARCH)
	var downloadURL, checksumsURL string
	for _, a := range release.Assets {
		switch a.Name {
		case wantName:
			downloadURL = a.BrowserDownloadURL
		case "checksums.txt":
			checksumsURL = a.BrowserDownloadURL
		}
	}
	if downloadURL == "" {
		errorJSON(c, http.StatusBadGateway, fmt.Sprintf("No asset found for %s", wantName))
		return
	}
	if checksumsURL == "" {
		errorJSON(c, http.StatusBadGateway, "Release has no checksums.txt — cannot verify integrity. Please update manually.")
		return
	}

	// 3. Download binary
	dlResp, err := http.Get(downloadURL)
	if err != nil {
		errorJSON(c, http.StatusBadGateway, "Failed to download binary")
		return
	}
	defer dlResp.Body.Close()

	if dlResp.StatusCode != http.StatusOK {
		errorJSON(c, http.StatusBadGateway, fmt.Sprintf("Download returned %d", dlResp.StatusCode))
		return
	}

	// 4. Write to temp file, computing SHA256 as we go
	exe, err := os.Executable()
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, "Cannot determine executable path")
		return
	}
	exe, err = filepath.EvalSymlinks(exe)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, "Cannot resolve executable path")
		return
	}

	tmpPath := exe + ".update"
	out, err := os.Create(tmpPath)
	if err != nil {
		errorJSON(c, http.StatusInternalServerError, "Cannot create temp file")
		return
	}

	hasher := sha256.New()
	if _, err := io.Copy(out, io.TeeReader(dlResp.Body, hasher)); err != nil {
		out.Close()
		os.Remove(tmpPath)
		errorJSON(c, http.StatusInternalServerError, "Failed to write binary")
		return
	}
	out.Close()
	actualHash := hex.EncodeToString(hasher.Sum(nil))

	// 5. Download and verify checksum
	expectedHash, err := fetchExpectedChecksum(checksumsURL, wantName)
	if err != nil {
		os.Remove(tmpPath)
		slog.Error("checksum fetch failed", "error", err)
		errorJSON(c, http.StatusBadGateway, "Failed to fetch checksums")
		return
	}
	if actualHash != expectedHash {
		os.Remove(tmpPath)
		slog.Error("checksum mismatch", "expected", expectedHash, "actual", actualHash)
		errorJSON(c, http.StatusBadGateway, "Checksum verification failed — downloaded binary may be corrupted")
		return
	}

	// 6. chmod +x and atomic rename
	if err := os.Chmod(tmpPath, 0o755); err != nil {
		os.Remove(tmpPath)
		errorJSON(c, http.StatusInternalServerError, "Failed to set permissions")
		return
	}
	if err := os.Rename(tmpPath, exe); err != nil {
		os.Remove(tmpPath)
		errorJSON(c, http.StatusInternalServerError, "Failed to replace binary")
		return
	}

	// 7. Respond success, then schedule self-termination
	successNoData(c)

	go func() {
		time.Sleep(500 * time.Millisecond)
		_ = syscall.Kill(syscall.Getpid(), syscall.SIGTERM)
	}()
}

// fetchExpectedChecksum downloads checksums.txt and returns the SHA256 hash
// for the given filename. The file format is "hash  filename" per line.
func fetchExpectedChecksum(checksumsURL, filename string) (string, error) {
	resp, err := http.Get(checksumsURL)
	if err != nil {
		return "", fmt.Errorf("download checksums: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("checksums returned %d", resp.StatusCode)
	}

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		// Format: "<hash>  <filename>" (two spaces, matching shasum output)
		parts := strings.Fields(scanner.Text())
		if len(parts) == 2 && parts[1] == filename {
			return parts[0], nil
		}
	}
	if err := scanner.Err(); err != nil {
		return "", fmt.Errorf("reading checksums: %w", err)
	}
	return "", fmt.Errorf("no checksum entry for %s", filename)
}
